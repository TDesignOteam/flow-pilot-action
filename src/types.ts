import type { RestEndpointMethodTypes } from '@octokit/plugin-rest-endpoint-methods'

export type PullRequestData = RestEndpointMethodTypes['pulls']['get']['response']['data']
export type PackagesChangelog = Record<string, string[]>
