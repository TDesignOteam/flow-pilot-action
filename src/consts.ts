export const SKIP_CHANGELOG_REG = /\[x\] 本条 PR 不需要纳入 Changelog/i
