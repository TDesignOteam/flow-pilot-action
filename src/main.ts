import { existsSync, readFileSync } from 'node:fs'
import path from 'node:path'
import { cwd } from 'node:process'
import { error, info } from '@actions/core'

export async function main() {
  const flowPilotPath = path.resolve(cwd(), '.flow-pilot.json')
  if (!existsSync(flowPilotPath)) {
    error('The file .flow-pilot.json does not exist.')
  }
  const flowPilotConfig = readFileSync(flowPilotPath, 'utf8')
  info(flowPilotConfig)
}
