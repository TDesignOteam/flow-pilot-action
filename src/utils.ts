import type { TokensList } from 'marked'
import { marked } from 'marked'

export function parseMarkdown(markdown: string): TokensList {
  return marked.lexer(markdown)
}

export function renderChangelog(markdown: string, pkgs: object) {
  const md = parseMarkdown(markdown)
  const logNode = parseMarkdown('### 📝 更新日志')[0]

  md.forEach((token) => {
    if (token.type === 'heading' && token.depth === logNode[0].depth && token.text === logNode[0].text) {
      token.text = `${pkgs[0]}: ${token.text}`
    }
  })
}
