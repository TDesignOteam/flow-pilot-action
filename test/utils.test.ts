import type { Tokens } from 'marked'
import { readFileSync } from 'node:fs'
import { describe, expect, it } from 'vitest'
import { parseMarkdown, renderChangelog, renderPackages } from '../src/utils'

describe('utils', () => {
  it('parseMarkdown', () => {
    const list = parseMarkdown('### 📝 更新日志')
    const data = list[0] as Tokens.Heading
    expect(list.length).toBe(1)
    expect(data.type).toBe('heading')
    expect(data.depth).toBe(3)
    expect(data.text).toBe('📝 更新日志')
  })

  it('renderPackages', () => {
    const packages = renderPackages('fixtures/repo1')
    expect(packages.length).toBe(2)
    expect(packages[0].relativeDir).toBe('packages/pkg-a')
    expect(packages[1].relativeDir).toBe('packages/pkg-c')
  })

  describe('renderChangelog', () => {
    it('repo1 pkg-b是私有包，不收集日志', () => {
      const packages = renderPackages('fixtures/repo1')
      expect(packages.length).toBe(2)
      expect(packages[0].relativeDir).toBe('packages/pkg-a')
      expect(packages[1].relativeDir).toBe('packages/pkg-c')
      const body = readFileSync('fixtures/pull_request_body/pr_body1.md', 'utf8').replaceAll('\n', '\r\n')
      const log = renderChangelog(body, packages.map(pkg => pkg.packageJson.name))
      expect(log).toMatchSnapshot()
    })
    it('repo2 无私有包', () => {
      const packages = renderPackages('fixtures/repo2')
      expect(packages.length).toBe(3)
      expect(packages[0].relativeDir).toBe('packages/pkg-a')
      expect(packages[1].relativeDir).toBe('packages/pkg-b')
      expect(packages[2].relativeDir).toBe('packages/pkg-c')
      const body = readFileSync('fixtures/pull_request_body/pr_body1.md', 'utf8').replaceAll('\n', '\r\n')
      const log = renderChangelog(body, packages.map(pkg => pkg.packageJson.name))
      expect(log).toMatchSnapshot()
    })
    it('本条 PR 不需要纳入 Changelog', () => {
      const packages = renderPackages('fixtures/repo1')

      const body = readFileSync('fixtures/pull_request_body/pr_body2.md', 'utf8').replaceAll('\n', '\r\n')
      const log = renderChangelog(body, packages.map(pkg => pkg.packageJson.name))
      expect(log).toBe(null)
    })
  })
})
