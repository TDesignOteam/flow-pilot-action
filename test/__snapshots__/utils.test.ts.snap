// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`utils > renderChangelog 1`] = `
{
  "pkg-a": [
    "feat(Button): New Component",
    "feat(Hook): add renderNodeJSX",
  ],
  "pkg-b": [
    "chore(utils): add \`isString\` function",
  ],
  "pkg-c": [
    "feat(Button): New Component",
  ],
}
`;
